hr_cae/models/hr_promotion.py

```diff
@@ -0,0 +1,123 @@
+        comodel_name="hr.employee", inverse_name="promotion_id", string="Employees"
+    )
+    applicant_ids = fields.One2many(
+        comodel_name="hr.applicant",
+        inverse_name="promotion_id",
+        string="Applicants",
+        domain=[("emp_id", "=", False)],
+    )
+    no_of_places_max = fields.Integer(string="Maximum Number of Places", copy=False)
+    no_of_places_taken = fields.Integer(
+        compute="_compute_no_of_places_taken", string="Places Taken", stre=True
+    )
+    no_of_places_available = fields.Integer(
+        compute="_compute_no_of_places_available", string="Places Available", store=True
+    )
+    no_of_applicants = fields.Integer(
+        compute="_compute_no_of_applicants", string="Applicants", store=True
+    )
+    attachment_number = fields.Integer(
+        compute="_compute_attachment_number", string="Number of Attachments"
+    )
+    attachment_ids = fields.One2many(
+        comodel_name="ir.attachment",
+        inverse_name="res_id",
+        string="Attachments",
+        domain=[("res_model", "=", "hr.promotion")],
+    )
+
+    @api.depends("employee_ids.active", "employee_ids.promotion_id")
+    def _compute_no_of_places_taken(self):
+        for promotion in self:
+            employees = self.env["hr.employee"].search(
+                [("active", "=", True), ("promotion_id", "=", promotion.id)]
+            )
+            promotion.no_of_places_taken = len(employees)
+
+    @api.depends("no_of_places_max", "no_of_places_taken")
+    def _compute_no_of_places_available(self):
+        for promotion in self:
+            if promotion.no_of_places_max:
+                promotion.no_of_places_available = (
+                    promotion.no_of_places_max - promotion.no_of_places_taken
+                )
+
+    @api.depends("applicant_ids.active", "applicant_ids.promotion_id")
+    def _compute_no_of_applicants(self):
+        for promotion in self:
+            applicants = self.env["hr.applicant"].search(
+                [("active", "=", True), ("promotion_id", "=", self.id)]
+            )
+            promotion.no_of_applicants = len(applicants)
+
+    @api.multi
+    @api.constrains("no_of_places_available")
+    def _constrain_no_of_places_available(self):
+        for promotion in self:
+            if promotion.no_of_places_available < 0 and promotion.no_of_places_max > 0:
+                raise ValidationError(
+                    _("Not enough places available in this promotion.")
+                )
+
+    @api.multi
+    def _compute_attachment_number(self):
+        read_group_res = self.env["ir.attachment"].read_group(
+            [("res_model", "=", "hr.promotion"), ("res_id", "in", self.ids)],
+            ["res_id"],
+            ["res_id"],
+        )
+        attach_data = {res["res_id"]: res["res_id_count"] for res in read_group_res}
+        for promotion in self:
+            promotion.attachment_number = attach_data.get(promotion.id, 0)
```

Comment:
Nice use of read_group. I'd use a better name for `attach_data` though.

Reply:
This was a clean copy from our code in `vertical_housing_cooperative`. I'd be interested to know how to improve and standardise this.